src/components/cylc/tree/index.js

```diff
@@ -0,0 +1,148 @@
+  const lookup = new Map()
+  // build hierarchy of cycle-point with zero or many families, and each family with zero or many other families
+  for (const familyProxy of workflow.familyProxies) {
+    const parent = familyProxy.firstParent
+    const parsedFamilyId = parseFamilyId(familyProxy.id)
+    if (!lookup.get(parsedFamilyId.cyclePoint)) {
+      // create cycle point node, using family's cycle point info
+      const cyclePointNode = {
+        __type: 'cyclepoint',
+        name: parsedFamilyId.cyclePoint,
+        children: [],
+        id: parsedFamilyId.cyclePoint,
+        state: ''
+      }
+      lookup.set(parsedFamilyId.cyclePoint, cyclePointNode)
+      // a cycle point must go directly under the workflow
+      workflow.children.push(cyclePointNode)
+    }
+
+    // skip the root family, which has null as its first parent
+    if (parent === null) {
+      continue
+    }
+    if (!lookup.get(familyProxy.id)) {
+      // create family proxy node for the tree
+      Object.assign(familyProxy, {
+        __type: 'family',
+        name: parsedFamilyId.family,
+        children: []
+      })
+      lookup.set(familyProxy.id, familyProxy)
+      // we should not add to the parent just yet, as it could a) not exist, or b) be the root family
+    }
+
+    const parsedParentId = parseFamilyId(parent.id)
+    if (parsedParentId.family === 'root') {
+      // if the parent is root, we use the cyclepoint as the parent
+      lookup.get(parsedFamilyId.cyclePoint).children.push(lookup.get(familyProxy.id))
+    } else {
+      if (!lookup.get(parent.id)) {
+        // construct preliminary parent node if not root
+        Object.assign(parent, {
+          __type: 'family',
+          name: parsedParentId.family,
+          children: []
+        })
+        lookup.set(parent.id, parent)
+      }
+      // attach the family node to its parent (cycle point, or another family, except root family)
+      lookup.get(parent.id).children.push(lookup.get(familyProxy.id))
+    }
+  }
+
+  workflow.children.sort((cyclepoint, anotherCyclepoint) => {
+    return cyclepoint.id.localeCompare(anotherCyclepoint.id)
+  })
+
+  // simply iterate through tasks, creating the nodes, then attach them to their parents using the lookup map
+  for (const taskProxy of workflow.taskProxies) {
+    Object.assign(taskProxy, {
+      __type: 'task',
+      name: taskProxy.task.name,
+      children: [],
+      expanded: false
+    })
+    const parsedTaskProxyId = parseFamilyId(taskProxy.firstParent.id)
+    if (parsedTaskProxyId.family === 'root') {
+      // if the parent is root, we must instead attach this node to the cyclepoint!
+      lookup.get(parsedTaskProxyId.cyclePoint).children.push(taskProxy)
+    } else {
+      lookup.get(taskProxy.firstParent.id).children.push(taskProxy)
+    }
+    for (const job of taskProxy.jobs) {
+      Object.assign(job, {
+        __type: 'job',
+        name: `#${job.submitNum}`,
+        latestMessage: taskProxy.latestMessage
```

Comment:
This is not exactly part of this issue. This code was moved from a `.vue` component to this `.js` file to simplify the component.

Can we move this for later? I think there was a discussion around this `latestMessage` before too. But I'd rather to leave this for later as we are currently seeing this latest message in the UI - I agree it's weird that we repeat the message of the task in all jobs.

Reply:
Fine, bump it to an issue if you like.

I think it would take some work to Cylc Flow for us to be able to provide `latestMessage` for each job.

Reply:
> I agree it's weird that we repeat the message of the task in all jobs.

Other way around: task proxy `latestMessage` is the latest message of the most recent job..

> I think it would take some work to Cylc Flow for us to be able to provide latestMessage for each job.

It’s already done, jobs have a list of all their messages... So we need only take the last in the list (or first if it’s sorted that way)...

Reply:
> It’s already done

Great!